internal/grpc/server/task_manager.go

```diff
@@ -89,7 +89,28 @@ func checkAuthorization(caller string, taskObj *taskmanager.Task) error {
 	return nil
 }
 
-// StreamTaskOutput
+// StreamTaskOutput outputs the task’s streams to the client by tracking its read offset.
 func (s *taskManagerServer) StreamTaskOutput(req *pb.StreamTaskOutputRequest, stream pb.TaskManager_StreamTaskOutputServer) error {
-	return status.Errorf(codes.Unimplemented, "StreamTaskOutput not implemented")
+	taskObj, err := s.taskManager.GetTask(stream.Context(), req.TaskId)
+	if err != nil {
+		return task.TaskErrorToGRPC(err)
+	}
+
+	caller := stream.Context().Value(basegrpc.ClientIDKey).(string)
+	if err = checkAuthorization(caller, taskObj); err != nil {
+		return err
+	}
+
+	// Create a writer that sends data over the gRPC stream
+	// stream.Send() will block if the client is slow to read the data
+	writer := func(data []byte) error {
+		return stream.Send(&pb.StreamTaskOutputResponse{Output: data})
+	}
+
+	err = s.taskManager.StreamTaskOutput(stream.Context(), req.TaskId, writer)
```

Comment:
What happens if the client closes the connection and no-new data is written to the output since then? 
Will the session terminate, will the goroutine become locked until a new write occurs?

Reply:
As the code stands now, the goroutine is locked until a new write occurs so the goroutines could pile up here if we had a bunch of clients that cancelled their context on a job that doesn't produce much output. I will add a recheck interval. Great catch @tigrato!

Reply:
What if you wake up the sync.Cond when a client wants to leave. Check context.AfterFunc .

That will also allow you to rewrite the streamer as an io.ReadCloser interface